### PR TITLE
[Modes] refactor modes to only use a stack in cpp

### PR DESCRIPTION
**Pull request by samdow**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #86458

Refactors the mode code to only have the C++ mode stack and not the "C++ mode" like we originally had. This also simplifies the mode logic in a number of places

### CODE REVIEW

File: test/test_overrides.py [hunk 1]
```diff
@@ -1220,6 +1220,7 @@ def __torch_function__(self, func, _, args=(), kwargs=None):
             with A("layer2"):
                 torch.empty([])
 
+        print(out)
```
Comment (zou3519) [inline test/test_overrides.py:1223]: erroneous print?

File: torch/csrc/utils/disable_torch_function.cpp [hunk 2]
```diff
@@ -221,7 +221,7 @@ inline bool has_torch_function_attr(PyObject* obj) {
 
 namespace torch {
 auto check_has_torch_function(PyObject* obj, bool ignore_mode) -> bool {
-  if (!ignore_mode && at::impl::PythonTorchFunctionTLS::get_mode())
+  if (!ignore_mode && at::impl::function_mode_enabled())
```
Comment (zou3519) [inline torch/csrc/utils/disable_torch_function.cpp:224]: nit: this is pre-existing, but function_mode_enabled() isn't descriptive to people who don't know the full name is "torch function mode" -- torch_function_mode_enabled() would be easier to read

**Review by zou3519 — APPROVED** [review]

File: torch/csrc/utils/disable_torch_function.cpp (hunk 2 above)
Comment (samdow) [reply torch/csrc/utils/disable_torch_function.cpp]: Agreed--since this is already a refactor, renamed this everywhere

**Comment by samdow** [comment]
pytorchbot merge -f "failure from cuda half. Unrelated to modes"

**Pull request closed without merging**